tools/make_patch.c

```diff
@@ -0,0 +1,476 @@
+
+struct Buffer {
+	size_t item_size;
+	size_t size;
+	size_t capacity;
+	char *data;
+};
+
+struct Patch {
+	unsigned int offset;
+	unsigned int size;
+};
+
+struct Symbol {
+	struct Symbol *next;
+	unsigned int address;
+	unsigned int offset;
+	char name[]; // VLA
+};
+
+struct Buffer *create_buffer(size_t item_size) {
+	struct Buffer *buffer = xmalloc(sizeof(*buffer));
+	buffer->item_size = item_size;
+	buffer->size = 0;
+	buffer->capacity = 0x10;
+	buffer->data = xmalloc(buffer->capacity * item_size);
+	return buffer;
+}
+
+void append_to_buffer(struct Buffer *buffer, const void *item) {
+	if (buffer->size >= buffer->capacity) {
+		buffer->capacity = (buffer->capacity + 1) * 2;
+		buffer->data = xrealloc(buffer->data, buffer->capacity * buffer->item_size);
+	}
+	memcpy(buffer->data + (buffer->size++ * buffer->item_size), item, buffer->item_size);
+}
+
+void free_buffer(struct Buffer *buffer) {
+	free(buffer->data);
+	free(buffer);
+}
+
+int get_address_type_limit(int address) {
+	return address >= 0x8000 && address < 0xa000 ? 0xa000 // VRAM
+	     : address >= 0xa000 && address < 0xc000 ? 0xc000 // SRAM
+	     : address >= 0xc000 && address < 0xf000 ? 0xd000 // WRAM
+	     : address >= 0xff80 && address < 0xffff ? 0x10000 // HRAM
+	     : 0x4000; // ROM
+}
+
+void append_symbol(struct Symbol **symbols, const char *name, int bank, int address) {
+	struct Symbol *symbol = xmalloc(sizeof(*symbol) + strlen(name) + 1);
+	symbol->address = address;
+	symbol->offset = bank > 0 ? address + (bank - 1) * get_address_type_limit(address) : address;
+	strcpy(symbol->name, name);
+	symbol->next = *symbols;
+	*symbols = symbol;
+}
+
+void free_symbols(struct Symbol *list) {
+	for (struct Symbol *next; list; list = next) {
+		next = list->next;
+		free(list);
+	}
+}
+
+const struct Symbol *find_symbol(const struct Symbol *symbols, const char *name) {
+	size_t namelen = strlen(name);
+	for (const struct Symbol *symbol = symbols; symbol; symbol = symbol->next) {
+		size_t symbolnamelen = strlen(symbol->name);
+		if (namelen > symbolnamelen) {
+			continue;
+		}
+		const char *symbolname = symbol->name;
+		if (name[0] == '.') {
+			// If `name` is a local label, compare it to the local part of `symbol->name`
+			symbolname += symbolnamelen - namelen;
+		}
+		if (!strcmp(symbolname, name)) {
+			return symbol;
+		}
+	}
+	error_exit("Error: Unknown symbol: %s\n", name);
+	return NULL;
+}
+
+const struct Symbol *find_symbol_cat(const struct Symbol *symbols, const char *prefix, const char *suffix) {
+	char *symname = xmalloc(strlen(prefix) + strlen(suffix) + 1);
+	strcpy(symname, prefix);
+	strcat(symname, suffix);
+	const struct Symbol *symbol = find_symbol(symbols, symname);
+	free(symname);
+	return symbol;
+}
+
+void parse_symbol_value(char *input, int *bank, int *address) {
+	char *buffer = xmalloc(strlen(input) + 1);
+	strcpy(buffer, input);
+
+	char *buffer2 = strchr(buffer, ':');
+	if (buffer2) {
+		// Parse symbol's bank:address
+		*buffer2++ = '\0';
```

Comment:
Instead of writing to `buffer2`, just check that `*endptr_bank == ':'`, i.e. that the first non-digit character was the `:` just `strchr`'d.

Not writing to `buffer2` means that you don't need to create a copy of the input string, either.

Reply:
We want to distinguish between failing to parse a `bank:address` string, and parsing a no-colon `constantvalue` string.

Reply:
Where is the distinction made?

Reply:
This is now moot because `parse_number` does its own error checking; and the input string can be modified anyway, so no need to malloc.